jenkins/shared-library/vars/defaultCiPipeline.groovy

```diff
@@ -0,0 +1,182 @@
+				choices: params.platform ?: 'ubuntu',
+				description: 'Run on specific platform'
+			booleanParam name: 'SHOULD_PUBLISH_IMAGE', description: 'true to publish image', defaultValue: false
+		}
+
+		agent {
+			label "${PLATFORM}-agent"
+		}
+
+		options {
+			ansiColor('css')
+			timestamps()
+		}
+
+		environment {
+			GITHUB_CREDENTIALS_ID = "${params.gitHubId}"
+			DOCKERHUB_CREDENTIALS_ID = 'docker-hub-token-symbolserverbot'
+			NPM_CREDENTIALS_ID = 'NPM_TOKEN_ID'
+			PYTHON_CREDENTIALS_ID = 'PYPI_TOKEN_ID'
+			DEV_BRANCH = 'dev'
+			RELEASE_BRANCH = 'main'
+			BUILD_SETUP_SCRIPT_FILEPATH = 'scripts/ci/setup_build.sh'
+			BUILD_SCRIPT_FILEPATH = 'scripts/ci/build.sh'
+			TEST_SETUP_SCRIPT_FILEPATH = 'scripts/ci/setup_tests.sh'
+			TEST_SCRIPT_FILEPATH = 'scripts/ci/test.sh'
+			LINTER_SCRIPT_FILEPATH = 'scripts/ci/linter.sh'
+		}
+
+		stages {
+			stage('CI pipeline') {
+				stages {
+					stage('display environment') {
+						steps {
+							println("Parameters: ${params}")
+							sh 'printenv'
+						}
+					}
+					stage('checkout') {
+						when {
+							expression { helper.isManualBuild() }
+						}
+						steps {
+							script {
+								gitCheckout(helper.resolveBranchName(env.MANUAL_GIT_BRANCH), env.GITHUB_CREDENTIALS_ID, env.GIT_URL)
+							}
+						}
+					}
+					stage('verify conventional commit message') {
+						steps {
+							verifyCommitMessage()
+						}
+					}
+					stage('lint') {
+						when { expression { return fileExists(resolvePath(packageRootPath, env.LINTER_SCRIPT_FILEPATH)) } }
+						steps {
+							runStepRelativeToPackageRoot packageRootPath, {
+								linter(env.LINTER_SCRIPT_FILEPATH)
+							}
+						}
+					}
+					stage('setup build') {
+						when {
+							expression {
+								return fileExists(resolvePath(packageRootPath, env.BUILD_SETUP_SCRIPT_FILEPATH))
+							}
+						}
+						steps {
+							runStepRelativeToPackageRoot packageRootPath, {
+								setupBuild(env.BUILD_SETUP_SCRIPT_FILEPATH)
+							}
+						}
+					}
+					stage('build code') {
+						steps {
+							runStepRelativeToPackageRoot packageRootPath, {
+								buildCode(env.BUILD_SCRIPT_FILEPATH)
+							}
+						}
+					}
+					stage('setup tests') {
+						when {
+							expression {
+								return fileExists(resolvePath(packageRootPath, env.TEST_SETUP_SCRIPT_FILEPATH))
+							}
+						}
+						steps {
+							runStepRelativeToPackageRoot packageRootPath, {
+								setupTests(env.TEST_SETUP_SCRIPT_FILEPATH)
+							}
+						}
+					}
+					stage('run tests') {
+						steps {
+							runStepRelativeToPackageRoot packageRootPath, {
+								runTests(env.TEST_SCRIPT_FILEPATH)
+							}
+						}
+					}
+				}
+			}
+			stage('CD pipeline') {
+				stages {
+					stage('publish RC') {
+						when {
+							anyOf {
+								branch env.DEV_BRANCH
+								allOf {
+									expression {
+										return env.SHOULD_PUBLISH_IMAGE.toBoolean()
+									}
+									not {
+										branch env.RELEASE_BRANCH
+									}
```

Comment:
why did you need this? aren't you checking it is DEV_BRANCH above?

Reply:
The ``anyOf`` is OR condition.  Its expands to ``( branch == dev)  or (should_publish_image && not main)``.  
While writing this makes me think of another scenario, should be able to release a RC build from main.  I will add another option which is cleaner